[JENKINS-54854] Added a warning when cron trigger spent more than a threshold (30s) in its execution

pjanouse opened the pull request:
See [JENKINS-54854](https://issues.jenkins-ci.org/browse/JENKINS-54854).

### Proposed changelog entries

* [JENKINS-54854] Added a warning when cron trigger spent more than a threshold (30s) in its execution

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

pjanouse committed:
[JENKINS-54854] Added a warning when cron trigger spent more than a threshold (30s) in its execution

pjanouse committed:
Added a test

olivergondza commented on core/src/main/java/hudson/triggers/Trigger.java:
```diff
@@ -275,7 +275,14 @@ public void run(AbstractProject p) {
+                                if ((end_time - begin_time) > CRON_THRESHOLD) {
+                                    LOGGER.log(Level.WARNING, "cron trigger " + t.getClass().getName() + ".run() triggered by {0} spent too much time ({1}) in its execution, " +
+                                            "other timers can be affected",
+                                            new Object[] {p, Util.getTimeSpanString(end_time - begin_time)});
```
We should either use plain old string concatenation or parameter replacement. Using both give us disadvantages from both worlds.

daniel-beck commented:
Is this something that should be exposed as an admin monitor even?

pjanouse committed:
Fix based on feedback

pjanouse committed:
Added Admin monitor

pjanouse commented:
I like the idea with admin monitor, so added it here.

pjanouse committed:
Cleanup

pjanouse committed:
Polishing

olivergondza commented on core/src/main/java/hudson/triggers/Trigger.java:
```diff
@@ -275,7 +275,18 @@ public void run(AbstractProject p) {
+                                            + " much time ({2}) in its execution, other timers can be affected",
+                                            new Object[] {t.getClass().getName(), p, Util.getTimeSpanString(end_time - begin_time)});
+                                    ADMIN_MONITOR.report(t.getClass().getName(),
+                                            String.format("Trigger %s.run() triggered by %s spent too much time "
```
Please assemble the message only once. There is no gain in using the logging late evaluation if we are doing `String.format` every time.

olivergondza commented on core/src/main/java/hudson/triggers/TriggerAdminMonitor.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Map;
+import java.util.logging.Logger;
+
+public class TriggerAdminMonitor extends AdministrativeMonitor {
```
This should be better restricted unless we want clients to consume this. Also `SlowTriggerAdminMonitor` feels more explicit.

olivergondza commented on core/src/main/java/hudson/triggers/TriggerAdminMonitor.java:
```diff
@@ -0,0 +1,79 @@
+        return new HashMap<>(errors);
+    }
+
+    public HttpResponse doClear() throws IOException {
```
Should require post method.

pjanouse replied:
Why? It doesn't come from a FORM, but link...

olivergondza replied:
That is how we intend that to be used. Though from CSRF protection perspective it can be abused just the same. The rule of thumb is the GET method requests should not modify server state.

pjanouse committed:
Fix according the feedback provided by Oliver

pjanouse committed:
Added missing renamed admin monitor class

pjanouse committed:
Better Web layout

oleg-nenashev commented on core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    @Nonnull
+    public String getDisplayName() {
+        return "Cron triggers monitor";
```
Make it localizable?

oleg-nenashev commented on core/src/main/java/hudson/triggers/Trigger.java:
```diff
@@ -334,4 +345,7 @@ public void run(AbstractProject p) {
     }
+
+    @Extension
+    public static final SlowTriggerAdminMonitor ADMIN_MONITOR = new SlowTriggerAdminMonitor();
```
Why don't you just put extension on the `SlowTriggerAdminMonitor` class? In such configuration the instance will be shared between unit tests, and I doubt it is a desirable behavior

oleg-nenashev commented on core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    public void report(@Nonnull String trigger, @Nonnull String msg) {
+        errors.put(trigger, new Value(msg));
```
🐛 reports may be stacking up infinitely and causing a memory leak until admin manually clears them

oleg-nenashev reviewed (changes requested):
@pjanouse thanks for the PR, it should be useful. Unfortunately it will rather cause a memory leak in the current implementation, because the size of the `reports` object is not limited. Unless admin manually clears the monitor, objects will be stacking up until NPE. Also, the UI may also be corrupted in the case of hundreds monitoring entries.

I recommend to limit the number of stored entries, e.g. to 10 latest events

pjanouse committed:
Limit stacked messages to 10 by default

pjanouse committed:
Improved replacement

oleg-nenashev reviewed (changes requested):
SlowTriggerAdminMonitor#errors is not thread safe. Since the size is limited, maybe it makes sense to use another collection type, e.g. just circular array

pjanouse committed:
Make SlowTriggerAdminMonitor#errors thread-safe

pjanouse commented:
> SlowTriggerAdminMonitor#errors is not thread safe. Since the size is limited, maybe it makes sense to use another collection type, e.g. just circular array

I've fixed this and make SlowTriggerAdminMonitor#errors thread-safe, thank you for pointing this out.

Also I though about to use other collection type (vector/array etc. with auto-override feature a.k.a. fixed capacity) along with thread-safe necessity, but realized it comes with more complexity on Jenkins code side without any benefit. Therefore I use just proper thread-safe collection type `ConcurrentHashMap` and synchronize access to it when necessary. I hope, I made it right and don't overlook sth.

oleg-nenashev reviewed (approved):
After a quick review I do not see anything obviously wrong in the last version.
Thanks @pjanouse for addressing the comments!

fcojfernandez reviewed (approved):
Nice PR!

oleg-nenashev commented:
It was around for 2+ weeks waiting for feedback, so I want to go forward and merge this PR which is definitely helpful for the instance monitoring

daniel-beck commented on core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java:
```diff
@@ -0,0 +1,121 @@
+    @Nonnull
+    public HttpResponse doClear() throws IOException {
+        Jenkins.getInstance().checkPermission(Jenkins.ADMINISTER);
+        LOGGER.info("Called clearing...");
```
```suggestion
```

daniel-beck commented on core/src/main/resources/hudson/triggers/Messages.properties:
```diff
@@ -35,3 +35,4 @@ TimerTrigger.the_specified_cron_tab_is_rare_or_impossible=This schedule will mat
   never (e.g. June 31), so this job may be triggered very rarely, if at all.
 Trigger.init=Initializing timer for triggers
 SCMTrigger.AdministrativeMonitorImpl.DisplayName=Too Many SCM Polling Threads
+SlowTriggerAdminMonitor.DisplayName="Cron triggers monitor"
```
```suggestion
SlowTriggerAdminMonitor.DisplayName=Cron triggers monitor
```

daniel-beck reviewed (commented):
Why does this not apply to synchronous SCM polling too?

daniel-beck reviewed (dismissed):
Requesting a fix to the display name:

> ![Screen Shot](https://user-images.githubusercontent.com/1831569/56849941-b5577b00-68fb-11e9-821f-016c20b0b8e8.png)

pjanouse committed:
Fixed as requested

pjanouse commented:
Fixed as requested by Daniel

pjanouse commented:
> Why does this not apply to synchronous SCM polling too?

Can be handled as another improvement separately?

daniel-beck commented on core/src/main/resources/hudson/triggers/Messages.properties:
```diff
@@ -35,3 +35,4 @@ TimerTrigger.the_specified_cron_tab_is_rare_or_impossible=This schedule will mat
   never (e.g. June 31), so this job may be triggered very rarely, if at all.
 Trigger.init=Initializing timer for triggers
 SCMTrigger.AdministrativeMonitorImpl.DisplayName=Too Many SCM Polling Threads
+SlowTriggerAdminMonitor.DisplayName=Cron triggers monitor
```
```suggestion
SlowTriggerAdminMonitor.DisplayName=Cron Triggers Monitor
```

All others are title-cased too.

pjanouse committed:
Admin monitor title changed

oleg-nenashev commented:
It was around for a while. I am going to go ahead and merge it towards the next weekly

jglick commented on test/src/test/java/jenkins/triggers/TriggerTest.java:
```diff
@@ -0,0 +1,118 @@
+            Thread.sleep(100);
+        }
+        j.waitUntilNoActivity();
+        j.interactiveBreak();
```
#4142

jglick commented on test/src/test/java/jenkins/triggers/TriggerTest.java:
```diff
@@ -0,0 +1,118 @@
+    public LoggerRule l = new LoggerRule();
+
+    @Test
+    public void testTimerSpentToMuchTime() throws Exception {
```
BTW there is a typo here

jsoref commented:
See #4176
